
{% extends 'base.html' %}
{% load crispy_forms_tags %}
{% load static%}


{% block header %}

<div class="jumbotron">
  <div class="container">
    <div class="row d-flex align-items-center">
      <div class="col-md-4">
        <img src="{% static 'media/ES.png' %}" alt="ExonSurfer Logo" class="img-fluid">
      </div>
      <div class="col-md-8">
        <h1 class="display-3"><b>ExonSurfer</b></h1>
        <p>A web-tool for exon-specific primer design</p>
      </div>
    </div>
  </div>
</div>

{% endblock header %}

{% block body %}
<div class="container">
  <!-- About ExonSurfer -->
  <div class="row gy-5">
    <div class="w-100" style="padding-top: 6%;"></div>
    <div class="col-lg">
      {% if messages %}
          <div>
              <ul>
              {% for message in messages %}
              <div class="alert alert-warning alert-dismissible fade show" role="alert">
                      <strong>{{message | safe}}</strong>
                      <button type="button" class="close" data-dismiss="alert" aria-label="Close">
                          <span aria-hidden="true">&times;</span>
                      </button>
              </div>
              {% endfor %}
              </ul>
          </div>  
        {% endif %}
        <p><h2><strong>About ExonSurfer</p></h2></strong>
      <p align="left">ExonSurfer is a web-tool for designing high-specificity exon junction binding transcript primers for qPCR, using the Ensembl database and the blastn algorithm. It combines all the steps of transcript-specific primer design, such as target selection to exclude other transcripts of the same gene, avoiding common polymorphic regions, and enables specificity and self-complementarity verification.</p>
      <p align="left">ExonSurfer simplifies the primer design process, while providing the user with a variety of options to customize the design to best suit their needs. In conclusion, this tool ensures high specificity and low non-specific binding of the primers, making ExonSurfer an essential tool for transcript-specific detection.</p>
    </div><!-- /.col-lg -->
  </div><!-- /.row -->
</div><!-- /.container -->

<!-- ... -->

<div class="container">
  <!-- Example row of columns -->
  <div class="row gy-5">
    <div class="w-100" style="padding-top: 6%;"></div>
      <div class="col-lg d-flex flex-column">
        <h2><strong>Design Primers</strong></h2>
        <p align="left">Create primers for your favorite species and gene using ExonSurfer's user-friendly interface and customizable parameters.</p>
       <!-- Design Primers Form -->
        <form method="post" enctype="multipart/form-data">
            {% csrf_token %}
            <!-- Form content -->

            <div class="form-group">
                <div class="form-group">
                    <p><label for="selectSpecies">Select species</label></p>
                    <select name="species" id="selectSpecies" class="form-control">
                        <option value="homo_sapiens">Homo sapiens</option>
                        <option value="mus_musculus">Mus musculus</option>
                        <option value="rattus_norvegicus">Rattus norvegicus</option>
                        <option value="drosophila_melanogaster">Drosophila melanogaster</option>
                        <option value="arabidopsis_thaliana">Arabidopsis thaliana</option>
                    </select>
                </div>
                <div class="form-group">
                    <label for="gene_field">Gene Symbol</label>
                    <input type="text" class="form-control" name="gene_field" placeholder="Enter gene symbol" required>
                </div>

                <div class="form-group">
                    <div class="form-check">
                        <input class="form-check-input" type="checkbox" name="use_masked_genomes" id="use_masked_genomes">
                        <label class="form-check-label" for="use_masked_genomes">
                            Select SNP-Masked Genome
                        </label>
                    </div>
                </div>
            </div>

            <!-- Submit Button -->
            <button type="submit" class="btn btn-primary" style="background-color: #355C8E; border-color:#355C8E">
                Select Transcript & Primer Parameters
            </button>
        </form>

   
      </div> <!-- /col-lg-6 -->

      <div class="col-lg d-flex flex-column">
        <h2><strong>Upload Custom Transcript Files</strong></h2>
        <p align="left">ExonSurfer supports custom transcript information. Upload your own FASTA or GenBank file if your gene or transcript isn't available in the Ensembl database.</p>
        <!-- Button 2 -->
        <div class="mt-auto"> <!-- Add "mt-auto" class to push the button to the bottom -->
          <a href="{% url 'primerblast_upload'%}" style="background-color: #355C8E; border-color:#355C8E" class="btn btn-primary btn-block">Upload</a>
        </div>
      </div> <!-- /col-lg-6 -->
  </div><!-- /.row -->
</div><!-- /.container -->

  <div class="container">
    <div class="row gy-5">
      <div class="w-100" style="padding-top: 6%;"></div>
      <div class="col-lg-12">
        <h2><strong>ExonSurfer: Code Availability and Usage</strong></h2>
        <p align="left">ExonSurfer is not only a versatile web-tool for generating primers but also a Python package accessible for integration into various research projects. For those interested in incorporating ExonSurfer into their workflow, the package is available at <a href="https://github.com/CrisRu95/ExonSurfer/tree/main">https://github.com/CrisRu95/ExonSurfer/tree/main</a>. Additionally, the web-tool interface can be built locally using Docker, offering greater flexibility and customization. Detailed instructions and all necessary resources for this setup can be found in the repository at <a href="https://github.com/pamonlan/ExonSurferWeb">https://github.com/pamonlan/ExonSurferWeb</a>.</p>
      </div><!-- /col-lg-4 -->
    </div><!-- /row -->
  </div><!-- /container -->


{% endblock body %}


{% block js %}

<script src="{% static 'index/js/form.js'%}"></script>

<script type="text/javascript">
  const selectSpecies = document.querySelector("#selectSpecies");
  const speciesInfo = document.querySelector("#speciesInfo");
  const speciesDivs = speciesInfo.children;

  selectSpecies.addEventListener("change", function() {
    const selectedValue = selectSpecies.value;

    for (let i = 0; i < speciesDivs.length; i++) {
      const speciesDiv = speciesDivs[i];
      if (speciesDiv.id === selectedValue) {
        speciesDiv.style.display = "block";
      } else {
        speciesDiv.style.display = "none";
      }
    }
  });

</script>
{% endblock js %}
